EventFilter/HcalRawToDigi/plugins/alpaka/HcalDigisProducerPortable.cc

```diff
@@ -0,0 +1,198 @@
+        uint16_t s1 = (0 << 7) | (static_cast<uint8_t>(hbhe.sample(2 * i + 1).adc()) & 0x7f);
+        uint16_t sample = (s1 << 8) | s0;
+        hf5_vi.data()[i + 1] = sample;
+      }
+    }
+    // copy hf5 to df5
+    alpaka::memcpy(event.queue(), df5_.buffer(), hf5_.const_buffer());
+    event.emplace(digisF5HBToken_, std::move(df5_));
+
+    if (qie11Digis.empty()) {
+      event.emplace(digisF01HEToken_);
+      event.emplace(digisF3HBToken_);
+
+    } else {
+      auto size_f1 = 0;
+      auto size_f3 = 0;
+
+      // count the size of the SOA;
+      for (unsigned int i = 0; i < qie11Digis.size(); i++) {
+        auto const digi = QIE11DataFrame{qie11Digis[i]};
+
+        if (digi.flavor() == 0 or digi.flavor() == 1) {
+          if (digi.detid().subdetId() == HcalEndcap) {
+            size_f1++;
+          }
+        } else if (digi.flavor() == 3) {
+          if (digi.detid().subdetId() == HcalBarrel) {
+            size_f3++;
+          }
+        }
+      }
+      auto const nsamples = qie11Digis.samples();
+      auto const stride01 = nsamples * QIE11DataFrame::WORDS_PER_SAMPLE + QIE11DataFrame::HEADER_WORDS;
+
+      // stack host memory in the queue
+      HostCollectionPhase1 hf1_(size_f1, event.queue());
+      HostCollectionPhase1 hf3_(size_f3, event.queue());
+
+      // device product
+      DeviceCollectionPhase1 df1_(size_f1, event.queue());
+      DeviceCollectionPhase1 df3_(size_f3, event.queue());
+
+      // set SoA_Scalar;
+      hf1_.view().stride() = stride01;
+      hf3_.view().stride() = stride01;
+
+      unsigned int i_f1 = 0;  //counters for f1 digis
+      unsigned int i_f3 = 0;  //counters for f3 digis
+
+      for (unsigned int i = 0; i < qie11Digis.size(); i++) {
+        auto const digi = QIE11DataFrame{qie11Digis[i]};
+        assert(digi.samples() == qie11Digis.samples() && "collection nsamples must equal per digi samples");
+
+        if (digi.flavor() == 0 or digi.flavor() == 1) {
+          if (digi.detid().subdetId() != HcalEndcap)
+            continue;
+          auto hf01_vi = hf1_.view()[i_f1];
+
+          hf01_vi.ids() = digi.detid().rawId();
+          for (int hw = 0; hw < QIE11DataFrame::HEADER_WORDS + digi.samples(); hw++) {
+            hf01_vi.data()[hw] = (qie11Digis[i][hw]);
+          }
+          i_f1++;
+        } else if (digi.flavor() == 3) {
+          if (digi.detid().subdetId() != HcalBarrel)
+            continue;
+          auto hf03_vi = hf3_.view()[i_f3];
+
+          hf03_vi.ids() = digi.detid().rawId();
+
+          for (int hw = 0; hw < QIE11DataFrame::HEADER_WORDS + digi.samples(); hw++) {
+            hf03_vi.data()[hw] = (qie11Digis[i][hw]);
+          }
+          i_f3++;
+        }
+      }
+
+      hf1_.view().size() = size_f1;
+      hf3_.view().size() = size_f3;
+
+      alpaka::memcpy(event.queue(), df1_.buffer(), hf1_.const_buffer());
+      alpaka::memcpy(event.queue(), df3_.buffer(), hf3_.const_buffer());
```

Comment:
Should we look for a way to avoid the copies, when the "device" is actually the cpu ?